src/streams.rs

```diff
@@ -0,0 +1,174 @@
+use crate::schemas;
+use crate::settings;
+use futures::channel::mpsc::{Receiver, Sender};
+use futures::sink::SinkExt;
+use futures::StreamExt;
+use log::*;
+use rdkafka::config::ClientConfig;
+use rdkafka::consumer::stream_consumer::StreamConsumer;
+use rdkafka::consumer::Consumer;
+use rdkafka::message::Message;
+use rdkafka::producer::{FutureProducer, FutureRecord};
+use std::collections::HashMap;
+use std::path::PathBuf;
+
+/// A struct to carry a message over a channel to be published into a destination Kafka topic.
+#[derive(Clone, Debug)]
+pub struct DispatchMessage {
+    destination: String,
+    payload: String,
+}
+
+pub async fn consume_topic(
+    kafka_settings: HashMap<String, String>,
+    topic: settings::Topic,
+    schema_root: PathBuf,
+    mut sender: Sender<DispatchMessage>,
+) -> Result<(), std::io::Error> {
+    let kafka_config = create_config(&kafka_settings);
+
+    let consumer: StreamConsumer = kafka_config.create().expect("Consumer creation failed");
+    consumer.subscribe(&[&topic.name]);
+
+    let mut stream = consumer.start();
+    let mut compiled_schemas = schemas::CompiledSchemas::new(schema_root);
+
+    debug!("Consuming from {}", topic.name);
+
+    while let Some(message) = stream.next().await {
+        match message {
+            Err(e) => {
+                error!("Failed to consume: {}", e);
+            }
+            Ok(message) => {
+                // TODO: might as well turn payload into an OwnedMessage since we need to copy it
+                // around a couple times anyways
+                let payload = match message.payload_view::<str>() {
+                    None => "",
+                    Some(Ok(buf)) => buf,
+                    Some(Err(e)) => {
+                        error!("Could not parse message: {:?}", e);
+                        ""
+                    }
+                };
+                debug!("Message consumed: {:?}", payload);
+
+                let value: Result<serde_json::Value, serde_json::error::Error> =
+                    serde_json::from_str(payload);
+                if value.is_err() {
+                    debug!("Non-JSON payload received: {:?}", payload);
+                    continue;
+                }
+                let value = value.unwrap();
+
+                let compiled_schema = compiled_schemas.get_compiled(&topic, &value);
```

Comment:
I don't believe the lazy population of compile schemas is concurrency-safe. Since this invocation may occur in multiple threads (since this is a spawned task), the underlying `HashMap` and `HashSet` inside of `compiled_schemas` may be accessed/mutated between multiple threads concurrently.

I suggest wrapping both with an async-std `RwLock` inside of `CompiledSchema`

Reply:
Agree, it definitely is not concurrency safe. But at the moment, there is not any concurrent usage since each consumer thread has its own instance and we are awaiting each element in the stream. I was planning to adjust for concurrency when I pick back up on the multi-topic subscription concept.

Reply:
Dag, I glossed over the `::new()` call a few lines up. Carry on.